### PR TITLE
Make updating replica atomic, balance plan idempotent

### PR DESCRIPTION
**Pull request by yah01**
Signed-off-by: yah01 <yang.cen@zilliz.com>
#17556 #17367

### CODE REVIEW

**Comment by yah01** [comment]
rerun ut

File: internal/querycoord/query_coord.go [hunk 1]
```diff
@@ -463,13 +470,14 @@ func (qc *QueryCoord) loadBalanceNodeLoop(ctx context.Context) {
...
 			qc.metricsCacheManager.InvalidateSystemInfoMetrics()
-			//TODO:: deal enqueue error
+
 			err := qc.scheduler.Enqueue(loadBalanceTask)
 			if err != nil {
 				log.Warn("failed to enqueue LoadBalance task into the scheduler",
 					zap.Int64("nodeID", node),
 					zap.Error(err))
 				qc.offlineNodesChan <- node
+				time.Sleep(LoadBalanceRetryAfter)
```
Comment (congqixia) [inline internal/querycoord/query_coord.go:480]: NIT: sleep or context.Done(), maybe use existing `waitWithContext` util func.

File: internal/querycoord/query_coord.go [hunk 2]
```diff
@@ -483,6 +491,7 @@ func (qc *QueryCoord) loadBalanceNodeLoop(ctx context.Context) {
...
```
Comment (congqixia) [inline internal/querycoord/query_coord.go:494]: NIT: same here. Context ignored.

File: internal/querycoord/query_coord.go [hunk 3]
```diff
@@ -341,16 +341,37 @@ func (qc *QueryCoord) watchNodeLoop() {
...
+	offlineNodes := make(typeutil.UniqueSet)
+	collections := qc.meta.showCollections()
+	for _, collection := range collections {
+		for _, replicaID := range collection.ReplicaIds {
+			replica, err := qc.meta.getReplicaByID(replicaID)
+			if err != nil {
+				log.Warn("failed to get replica",
+					zap.Int64("replicaID", replicaID),
+					zap.Error(err))
+				continue
```
Comment (congqixia) [inline internal/querycoord/query_coord.go:361]: NIT: what may cause replica not found? Shall we remove it from collection meta?

**Review by congqixia — APPROVED** [review]
/lgtm

File: internal/querycoord/query_coord.go (hunk 1 above)
Comment (yah01) [reply internal/querycoord/query_coord.go:480]: This line is in a loop that select ctx.Done()

**Pull request merged**